sdmx30-infomodel/src/main/java/com/epam/jsdmx/infomodel/sdmx30/VersionReference.java

```diff
@@ -1,58 +1,140 @@
-    private static final Pattern PATCH_WILDCARD_PATTERN = Pattern.compile("(0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\+(-\\w+)?");
-
-    private static final List<Pair<Pattern, Function<String, VersionReference>>> CREATORS_BY_PATTERN = List.of(
-        Pair.of(Version.PATTERN, (String s) -> new VersionReference(Version.createFromString(s), WildcardScope.NONE)),
-        Pair.of(MINOR_WILDCARD_PATTERN, (String s) -> new VersionReference(Version.createFromString(removeWildcard(s)), WildcardScope.MINOR)),
-        Pair.of(MAJOR_WILDCARD_PATTERN, (String s) -> new VersionReference(Version.createFromString(removeWildcard(s)), WildcardScope.MAJOR)),
-        Pair.of(PATCH_WILDCARD_PATTERN, (String s) -> new VersionReference(Version.createFromString(removeWildcard(s)), WildcardScope.PATCH))
-    );
-
     private VersionReference(Version version, WildcardScope scope) {
         super(version, scope);
     }
 
-    private static String removeWildcard(String s) {
-        return s.replaceAll("\\+", "");
-    }
-
     /**
      * Creates a {@link VersionReference} from a string representation.
+     *
      * @param version valid string representation of a referenced version
      */
     public static VersionReference createFromString(String version) {
-        return createFromString(CREATORS_BY_PATTERN, version);
+        try {
+            return parse(version);
+        } catch (IllegalArgumentException | IndexOutOfBoundsException e) {
+            throw new IllegalArgumentException("Invalid version: " + version);
+        }
+    }
+
+    private static VersionReference parse(String version) {
+        WildcardScope scope = NONE;
+        int majorUntil = -1;
+        int minorFrom = -1;
+        int minorUntil = -1;
+        int patchFrom = -1;
+        int patchUntil = -1;
+        int extensionFrom = -1;
+        Position p = Position.MAJOR;
+        final int inputLength = version.length();
+        for (int i = 0; i < inputLength; i++) {
+            char c = version.charAt(i);
+            if (c > 47 & c < 58) {
+                switch (p) {
+                    case MAJOR:
+                        majorUntil = i;
+                        break;
+                    case MINOR:
+                        minorUntil = i;
+                        break;
+                    case PATCH:
+                        patchUntil = i;
+                        break;
+                    default:
+                        throw new IllegalArgumentException();
+                }
+            } else if (c == '+') {
+                switch (p) {
+                    case MAJOR:
+                        scope = MAJOR;
+                        break;
+                    case MINOR:
+                        scope = MINOR;
+                        break;
+                    case PATCH:
+                        scope = WildcardScope.PATCH;
```

Comment:
We should remove WildcardScope. or add it to other cases in current switch to not confuse with Position enum

Reply:
Unfortunately, they are not the same, and I need `position` to keep track of the current placement in the version being read, and, at the same time, once we hit `+` we should update scope value to track the type of the wildcard. I've done some renaming to improve the situation.